honeybee_radiance/lightsource/sky/cie.py

```diff
@@ -0,0 +1,291 @@
+            ground_reflectance: Average ground reflectance (Default: 0.2).
+        """
+        assert isinstance(location, Location), \
+            'location must be from type Location not {}'.format(type(location))
+        return cls.from_lat_long(
+            location.latitude, location.longitude, location.time_zone, month, day, hour,
+            sky_type, north_angle, ground_reflectance)
+
+    @classmethod
+    def from_dict(cls, input_dict):
+        """Create the sky from a dictionary.
+
+        Args:
+            input_dict: A python dictionary in the following format
+
+        .. code-block:: python
+
+                {
+                'altitude': 0.0,
+                'azimuth': 0.0,
+                'sky_type': 0,
+                'ground_reflectance': 0.2,
+                'ground_hemisphere': {},  # see ground.Ground class [optional],
+                'sky_hemisphere': {}  # see hemisphere.Hemisphere class [optional]
+                }
+        """
+        sky = cls(
+            input_dict['altitude'],
+            input_dict['azimuth'],
+            input_dict['sky_type'],
+            input_dict['ground_reflectance']
+        )
+
+        if 'ground_hemisphere' in input_dict:
+            sky._ground_hemisphere = Ground.from_dict(input_dict['ground_hemisphere'])
+
+        if 'sky_hemisphere' in input_dict:
+            sky._sky_hemisphere = Hemisphere.from_dict(input_dict['sky_hemisphere'])
+
+        return sky
+
+    # TODO: add support for additional parameters
+    # TODO: add gensky to radiance-command and use it for validating inputs
+    def to_radiance(self):
+        """Return radiance definition as a string."""
+        command = '!gensky -ang %.6f %.6f %s -g %.3f' % (
+            self.altitude, self.azimuth - 180.0, self.sky_type_radiance,
+            self.ground_reflectance
+        )
+
+        return '%s\n\n%s\n\n%s\n' % (
+            command, self.sky_hemisphere, self.ground_hemisphere
+        )
+
+    def to_dict(self):
+        """Translate sky to a dictionary."""
+        return {
+            'altitude': self.altitude,
+            'azimuth': self.azimuth,
+            'sky_type': self.sky_type,
+            'ground_reflectance': self.ground_reflectance,
+            'ground_hemisphere': self.ground_hemisphere.to_dict(),
+            'sky_hemisphere': self.sky_hemisphere.to_dict()
+        }
+
+    def to_file(self, folder, name=None, mkdir=False):
+        """Write sky hemisphere to a sky_hemisphere.rad Radiance file.
+
+        Returns:
+            Full path to the newly created file.
+        """
+        content = self.to_radiance()
+        name = typing.valid_string(name) if name \
+            else '%.3f_%.3f_%s.sky' % (
+                self.altitude, self.azimuth,
+                self.sky_type_human_readable.replace(' ', '_').lower()
+            )
+        return futil.write_to_file_by_name(folder, name, content, mkdir)
+
+    def __eq__(self, value):
```

Comment:
> ... I think most other developers would agree with me.

It probably depends on the developer and how much they have played with hash! 😀

The general rule is that mutable objects in python must not be used as dictionary keys unless you can write a hash which never changes during the lifetime of the object. That's not the case here and is not the case in most of the objects that we have in Ladybug Tools libraries.

I know you have implemented this to several classes and I remember you mainly referred to better performance as the reason. I trusted that you have tested it on your side otherwise the logic for creating a hash based on mutable values will not hold up to your expectations. Here is an example:

```python
class A:

    def __init__(self, a, b):
        self.a = a
        self.b = b

    def __key(self):
        return self.a, self.b

    def __hash__(self):
        return hash(self.__key())

    def __eq__(self, other):
        return isinstance(other, A) and self.__key() == other.__key()


a = A(1, 2)

d = {a: 'hi'}

a.b = 5  # oops!

print(d[a])
# you will get a KeyError

```
Here is a longer post: https://inventwithpython.com/blog/2019/02/01/hashable-objects-must-be-immutable/

Reply:
That is a really good point. Dictionary keys were a really bad example and you are right that we shouldn't use mutable or lockable objects for dictionary keys. Sets should still be ok for locked objects, though.

I also forgot that these skies are not lockable like all of the other places where we've implemented aligned hash and eq methods. So I'm cool with your current implementation now that I realized this.